docs: HA distributed query decisions (#8817)
spiceai/spiceai: docs/decisions/006-ha-distributed-query.md

## Assumptions

1. Network topology:
  a. allows executors to connect to schedulers.
  b. may or may not allow executors to connect to other executors.
  c. allows schedulers to connect to other schedulers.

> 2. Orchestration system is responsible for maintaining minimum cluster requirements.

Reviewer: ```suggestion
2. Orchestration system is responsible for maintaining minimum host availability.
```

## Options

### Shared state and conflict resolution

How schedulers store and coordinate shared cluster state, and how conflicting updates are resolved. Cluster state includes scheduler/executor membership and in-flight job/task metadata.

* Object store shared state with conditional writes.
* External coordination services (e.g., etcd, ZooKeeper, Redis) for shared state and leader election.
* SQL database or distributed KV store for job and scheduler state.
* Object store with separate lock service instead of conditional writes.

### Discovery and connectivity

How schedulers and executors find each other and which side initiates and maintains network connections.

* Scheduler discovery via shared state registration; executor discovery via logical connections; executor-initiated scheduler connections.
* Static configuration (predefined scheduler and executor lists).
* Service discovery via DNS/SRV or platform-native registry.
* Bidirectional scheduler-executor connections with scheduler-initiated calls.

### Cluster composition

> The shape of the cluster and how schedulers and executors are arranged and shared across the cluster.

Reviewer: Is shape a techical term? Isn't this just topology?

* Multiple active schedulers with executors connected to all schedulers; executors may be shared.
* Active/passive scheduler with failover.
* Dedicated executors per scheduler (no sharing).
* Executors connected to a subset of schedulers instead of all.

### Execution and shuffle behavior

Execution guarantees and where shuffle/intermediate data is stored during query execution.

* Exactly-once job execution (stages/tasks at-least once); shuffle in object store or executor local storage.
* At-least-once job execution with client-level deduplication.
* Shuffle state always in object store (no local executor shuffle).
* External shuffle service separate from executors.

## Decision

* **Shared state and conflict resolution**

>   * Schedulers share state that is stored in object store.

Reviewer: ```suggestion
  * Schedulers share state and that state is stored in object store.
```

>   * Job state is stored in shared state.

Reviewer: ```suggestion
  * Job state is part of the shared state.
```
